Windows Xenstore Support

JKDingwall opened the pull request:
Hi,

This branch allows PyXS to work on Windows in combination with the GPLPV drivers.  The tested environment is:
Windows 2008R2
GPLPV 0.11.0.3
Python 2.7.6
pywin32-218.win-amd64-py2.7

The modified code was successfully tested on Windows and Linux with:

``` python
import os
import pyxs

if os.name in ["nt"]:
    xs = pyxs.Client(None)
elif os.name in ["posix"]:
    xs = pyxs.Client(xen_bus_path="/proc/xen/xenbus")

print xs.ls("/tools/data/vminfo")
print xs.read("vm")
```

JKDingwall committed:
Add some stubs for Windows xenstore support

JKDingwall committed:
Implement osnmread, osnmwrite, osnmopen, osnmclose for os.name = ["nt"]

JKDingwall committed:
Use the Windows API to find the device path for the xenbus device.

JKDingwall committed:
Some little missing pieces and fixes.  Successfully tested a xenstore list command!y

superbobry commented on pyxs/connection.py:
```diff
@@ -20,11 +20,24 @@
 
+if os.name in ["nt"]:
+    import ctypes
+    from ctypes.wintypes import HANDLE
```
I think importing `ctypes.wintypes` qualified might improve readability:

``` python
from ctypes import wintypes

foo(wintypes.BOOL)
```

JKDingwall replied:
A copy of the code from the serial example, as with the other piece below it could be changed if preferred.

superbobry commented on pyxs/connection.py:
```diff
@@ -105,11 +118,15 @@ def recv(self):
             return Packet(op, payload, rq_id, tx_id)
 
 
+    def __del__(self):
```
I'm not sure about this, since it encourages bad code: the user should not rely on the gc to close the handle. For the exact same reasons file handles are not closed on gc in СPython 3, see [link](http://stackoverflow.com/questions/7395542/is-explicitly-closing-files-important).

JKDingwall replied:
856010674434b31be9980b5aa7198a7d7be3bdeb removes the **del**()

superbobry commented on pyxs/connection.py:
```diff
@@ -180,7 +198,186 @@ def connect(self):
+        # http://pydoc.net/Python/pyserial/2.6/serial.tools.list_ports_windows/.
+        # The equivalent C from The GPLPV driver source is:
+        """\
+DEFINE_GUID(GUID_XENBUS_IFACE, 0x14ce175a, 0x3ee2, 0x4fae, 0x92, 0x52, 0x0, 0xdb, 0xd8, 0x4f, 0x1, 0x8e);
```
Can we reference this code instead of inlining it?

JKDingwall replied:
41ac0988679c25cc59159f39e668b24f8a66a9d5 removes the inline C.

superbobry commented on pyxs/connection.py:
```diff
@@ -180,7 +198,186 @@ def connect(self):
+                    self.Data1,
+                    self.Data2,
+                    self.Data3,
+                    ''.join(["%02x" % d for d in self.Data4[:2]]),
```
You don't need a list comprehension here, using a generator should be fine:

``` python
>>> "".join("%02x" % x for x in range(4))
'00010203'
```

superbobry replied:
Also, are you using Captialized attributes just to match the C code above?

JKDingwall replied:
These parts were copied from the serial example code rather than a match for the C.  I could change this if you prefer.

JKDingwall replied:
I think the capitalization is to match the types used in Windows API, e.g. http://msdn.microsoft.com/en-us/library/windows/hardware/ff551015%28v=vs.85%29.aspx.

superbobry commented on pyxs/connection.py:
```diff
@@ -180,7 +198,186 @@ def connect(self):
+        sdid.cbSize = ctypes.sizeof(sdid)
+        if not SetupDiEnumDeviceInterfaces(handle, NULL, ctypes.byref(GUID_XENBUS_IFACE), 0, ctypes.byref(sdid)):
+            if ctypes.GetLastError() != ERROR_NO_MORE_ITEMS:
+                    raise ctypes.WinError()
```
I think you should raise a subclass of `PyXsError` here.

JKDingwall replied:
I have added WindowsDriverError as a subclass of PyXSError.

JKDingwall committed:
Add WindowsDriverError exception and use it when a failure to find the xenbus device occurs.

JKDingwall committed:
Merge branch 'windows' of github.com:JKDingwall/pyxs into windows

Conflicts:
	pyxs/connection.py

JKDingwall committed:
Don't provide the C code the work is derived from inline, reference by URL to the GPLPV mercurial tree.

JKDingwall committed:
Remove __del__()

JKDingwall committed:
ctypes.POINTER() allocates memory which does not get reclaimed by gc to over time this leaks memory if multiple
instances of the connection class get created.  To avoid this once the device path has been discovered cache
the value and resuse that for every future invocation.

JKDingwall committed:
Avoid trailing NULL in value on write

JKDingwall committed:
Merge remote-tracking branch 'origin/master' into windows

royjenkins committed:
Add WMI support for Windows Server 2012 with Xen Project PV drivers

royjenkins committed:
Raise PyXSError properly

royjenkins committed:
Don't call connect() uneccesarily

royjenkins committed:
Stop session being created as a list

royjenkins committed:
Never allow _wmiSession to be null

royjenkins committed:
Retry Win32OperatingSystem WMI request if it fails

royjenkins committed:
Add additional retries with half-second sleeps around WMI requests; they can be a little unreliable

royjenkins committed:
Increase wait period on retry

royjenkins committed:
Many more retries

royjenkins committed:
< not >

JKDingwall committed:
except: is bad practice, don't do it

JKDingwall committed:
whitespace cleanup

JKDingwall committed:
Mark the xenbus connection according to the pv driver project not the windows version.

JKDingwall committed:
Try to decide on the windows xenbus connection class by looking for the project drivers rather than guess by operating system version.

JKDingwall committed:
remove unused import.

JKDingwall committed:
remove unused pythoncom import

JKDingwall committed:
workaround for http://lists.xenproject.org/archives/html/win-pv-devel/2016-02/msg00005.html

superbobry commented:
Hi, sorry for taking so long to get to this. 

Is there a spec. for accessing XenStore on Windows? I do hope there's a more consistent (with Linux) way of approaching the issue.

JKDingwall commented:
The fork on my account now has support for both the GPLPV and the WinPV driver projects.  The original support for GPLPV uses the Windows API to open a connection to the xenbus which is then used exactly the same as under Linux.  It is quite complicated because of how the xenbus device is identified and opened but essentially the same code as done in the block driver.  The WinPV project exposes xenstore through WMI so we have created a bridge to that in PyXS.  There are currently some issues in how the WinPV driver handles empty strings in xenstore as it returns them as None rather than "".  This has been raised on the WinPV mailing list.

superbobry commented:
This sounds very exciting. However, I'm not sure I could maintain this code as part of `pyxs`, since I'm not at all familiar with the Windows side of things. 

Here's what I propose instead:

* Convert a fork into a separate library, e.g. `winpyxs`.
* The library will extend ``pyxs`` with two connection classes: `XenBusConnectionWinWINPV` and `XenBusConnectionWinGPLPV` and provide Windows-specific wrappers for ``pyxs.monitor`` and ``pyxs.Client``. 

What do you think?

P.S. [As of 0.4.0] [router] you need a `pyxs.Router` to plug a custom connection class into `pyxs.Client`:

```python
router = Router(CustomConnectionClass())
with pyxs.Client(router=router) as c:
    # ...
```

[router]: http://pyxs.readthedocs.org/en/stable/api.html#pyxs.client.Client

superbobry commented:
I did a quick googling and failed to locate any developer documentation for GPLPV and WinPV drivers. How did you work out how to use these?

JKDingwall committed:
prefer "/dev/xen/xenbus" to "/proc/xen/xenbus" on Linux

JKDingwall committed:
prefer XenBusConnection over UnixSocketConnection on posix platforms

JKDingwall committed:
Rename DEBUG to CONTROL to match xen-4.9.  Don't send an empty payload with CONTROL.

JKDingwall committed:
Use noop control packet while waiting

The check operation essentially does an fsck of the xenstore tdb database which is expensive.  A small xenstore patch can be used to introduce a noop control function and it is preferable to use that.  No operation should be necessary here but the implementation of the wait() is wrong in this branch.  The upstream release handles this correctly but the change would be quite intrusive.

joelnb commented:
I had a go at implementing this as suggested in a separate module (entirely based on the work here) and was able to get a version which works with the current version of pyxs. There are some rough edges still and it's possible that some features are not working correctly but reading keys/listing directories is definitely fully functional. There's no documentation or tests currently but I plan to sort that out, hopefully over the coming week!

The repo is located here: https://github.com/joelnb/win-pyxs
And I pushed it to PyPI for easier installation (in case anyone wants to test): https://pypi.org/project/win-pyxs/
